src/beamlime/applications/handlers.py

```diff
@@ -0,0 +1,114 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+import pathlib
+from dataclasses import dataclass
+from typing import Any, NewType
+
+import plopp as pp
+import scipp as sc
+
+from beamlime.logging import BeamlimeLogger
+
+from ._workflow import Events, FirstPulseTime, Histogrammed, WorkflowPipeline
+from .base import BeamlimeMessage, HandlerInterface
+
+
+@dataclass
+class HistogramUpdated(BeamlimeMessage):
+    content: pp.graphics.basefig.BaseFig
+
+
+@dataclass
+class RawDataSent(BeamlimeMessage):
+    content: Events
+
+
+class DataReductionHandler(HandlerInterface):
+    """Data reduction handler to process the raw data and update the histogram.
+
+    It receives a list of events, and reduces them into a histogram.
+    It also triggers the update of a plot stream node.
+    """
+
+    pipeline: WorkflowPipeline
```

Comment:
Where does `pipeline` get initialised? I don't see it in the class.

Reply:
It's assembled by ``Factory``.
The provider is here:
https://github.com/scipp/beamlime/blob/c0264d838e02e73ba977644bc31a0dde324c77eb/src/beamlime/applications/_workflow.py#L134

Reply:
Can you point to the place where the attribute gets set?

Does this mean that the handler is in an invalid state until it gets patched up by the factory later?

Reply:
Ah...! I understood what you meant. It seems wrong...
I'll move it to the constructor instead.

> Does this mean that the handler is in an invalid state until it gets patched up by the factory later?

And yes...! It's invalid if it's not patched... I'll change...